### PR TITLE
Import source classes from rspec-core

### PR DESCRIPTION
**Pull request by yujinakayama**
For https://github.com/rspec/rspec-expectations/pull/987.

This corresponds to https://github.com/rspec/rspec-core/pull/2420.

### CODE REVIEW

File: lib/rspec/support/source/node.rb [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+          :word_new,
+          :words_add,
+          :words_new,
+          :xstring_add,
+          :xstring_literal,
+          :xstring_new,
+          :yield,
+          :yield0,
+          :zsuper
+        ]
```
Comment (myronmarston) [inline lib/rspec/support/source/node.rb]: I worry that this array will get out of sync with the ripper constant.  It looks like it's only needed to dynamically define all the `*_type?` methods.  Instead, WDYT about defining  a single `of_type?(event)` method or just do manual `==` comparisons?  Then you wouldn't need this constant, this class would be smaller/simpler, and we wouldn't have to worry about the constant going out of sync with Ripper's.

On a side note: the rspec-core `Node` class did not need this, so we do we need it now?

File: lib/rspec/support/source/node.rb [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+      private
+
+        def raw_args
+          sexp[1..-1] || []
+        end
+      end
+
+      # @private
+      class ExpressionSequenceNode < Node
```
Comment (myronmarston) [inline lib/rspec/support/source/node.rb]: I noticed rspec-core had `GroupNode` vs here you have `ExpressionSequenceNode`.  Can you explain why it needed to change?

BTW, a comment in the source explaining the need for this subclass would be helpful to our future selves, I think.

File: lib/rspec/support/source/token.rb [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+          :symbeg,
+          :symbols_beg,
+          :tlambda,
+          :tlambeg,
+          :tstring_beg,
+          :tstring_content,
+          :tstring_end,
+          :words_beg,
+          :words_sep
+        ].freeze
```
Comment (myronmarston) [inline lib/rspec/support/source/token.rb]: Here too I'd like to see us get away from a constant we have to keep in sync with ripper...and I'm also curious as to why we need it here but did not in rspec-core.

File: lib/rspec/support/source/location.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+        def self.location?(array)
+          array.is_a?(Array) && array.size == 2 && array.all? { |e| e.is_a?(Integer) }
+        end
+
+        def <=>(other)
+          line_comparison = (line <=> other.line)
+          return line_comparison unless line_comparison == 0
+          column <=> other.column
+        end
```
Comment (yujinakayama) [inline lib/rspec/support/source/location.rb]: It's needed here https://github.com/rspec/rspec-expectations/pull/987/commits/c074ca900790336204ae3e18e88c03ebf337ce24#diff-9b277082395427eb0159d15a3204096bR176.

File: lib/rspec/support/source/node.rb (hunk 1 above)
Comment (yujinakayama) [inline lib/rspec/support/source/node.rb]: This was for convenience:

* `node.select { |node| node.type == :fcall }` can be described as `node.select(&:fcall_type?)`
* We can notice if there's a typo in type name (`node.fcalll_type? # NoMethodError`)

> WDYT about defining a single `of_type?(event)` method or just do manual `==` comparisons?

Will do.

File: lib/rspec/support/source/node.rb (hunk 2 above)
Comment (yujinakayama) [inline lib/rspec/support/source/node.rb]: It's for clarity, since _group_ was too generic word.

Comment (yujinakayama) [inline lib/rspec/support/source/node.rb]: > BTW, a comment in the source explaining the need for this subclass would be helpful to our future selves, I think.

Done

**Pull request merged**